plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -1392,6 +1398,42 @@ public void setColumnComment(ConnectorSession session, ConnectorTableHandle tabl
         return viewToken;
     }
 
+    @Override
+    public Optional<CatalogSchemaTableName> redirectTable(ConnectorSession session, SchemaTableName tableName)
+    {
+        return catalog.redirectTable(session, tableName);
+    }
+
+    @SuppressWarnings("TryWithIdenticalCatches")
+    @Override
+    public Stream<TableColumnsMetadata> streamTableColumns(ConnectorSession session, SchemaTablePrefix prefix)
+    {
+        requireNonNull(prefix, "prefix is null");
+        return catalog.listTables(session, prefix).stream().flatMap(tableName -> {
+            try {
+                if (redirectTable(session, tableName).isPresent()) {
+                    return Stream.of(TableColumnsMetadata.forRedirectedTable(tableName));
```

Comment:
Is there a condition where they might have the same schema.tablename in both the iceberg and hive catalogs, and redirecting to hive as the first option wold result in unexpected behavior? Or is this function gated on the iceberg catalog acknowledging it has the same megastore as the hive catalog?

Reply:
h2. Tentative answer 1
If I understand what you are asking, there is a redirection cycle detection mechanism in :

`io.trino.metadata.MetadataManager#getRedirectedTableName`

https://github.com/trinodb/trino/blob/c13f9523ba88f30799c6e00e9c8ccc217f42e3fe/core/trino-main/src/main/java/io/trino/metadata/MetadataManager.java#L1544

h2. Tentative answer 2

HMS and Glue both store metadata about the tables and the tables are internally identified by:
- schema name
- table name

So, AFAIK it is not possible to have a hive table `default.test1` in hive  and  also an iceberg table `default.test1`.